pkg/trace/api/pipeline_stats.go

```diff
@@ -0,0 +1,104 @@
+)
+
+const (
+	pipelineStatsURLSuffix = "/api/v0.1/pipeline_stats"
+)
+
+// pipelineStatsEndpoint returns the pipeline intake url and the corresponding API key.
+func pipelineStatsEndpoint(cfg *config.AgentConfig) (url *url.URL, apiKey string, err error) {
+	if e := cfg.Endpoints; len(e) == 0 || e[0].Host == "" || e[0].APIKey == "" {
+		return nil, "", errors.New("config was not properly validated")
+	}
+	urlStr := cfg.Endpoints[0].Host + pipelineStatsURLSuffix
+	url, err = url.Parse(urlStr)
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing pipeline stats intake URL %s: %v", urlStr, err)
+	}
+	return url, cfg.Endpoints[0].APIKey, nil
+}
+
+// pipelineStatsProxyHandler returns a new HTTP handler which will proxy requests to the pipeline stats intake.
+func (r *HTTPReceiver) pipelineStatsProxyHandler() http.Handler {
+	target, key, err := pipelineStatsEndpoint(r.conf)
+	if err != nil {
+		log.Errorf("Failed to start pipeline stats proxy handler: %v", err)
+		return pipelineStatsErrorHandler(err)
+	}
+	tags := fmt.Sprintf("host:%s,default_env:%s,agent_version:%s", r.conf.Hostname, r.conf.DefaultEnv, info.Version)
+	if orch := r.conf.FargateOrchestrator; orch != fargate.Unknown {
+		tag := fmt.Sprintf("orchestrator:fargate_%s", strings.ToLower(string(orch)))
+		tags = tags + "," + tag
+	}
+	return newPipelineStatsProxy(r.conf.NewHTTPTransport(), target, key, tags)
+}
+
+func pipelineStatsErrorHandler(err error) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		msg := fmt.Sprintf("Pipeline stats forwarder is OFF: %v", err)
+		http.Error(w, msg, http.StatusInternalServerError)
+	})
+}
+
+// newPipelineStatsProxy creates an http.ReverseProxy which forwards requests to the pipeline stats intake.
+// The tags will be added as a header to all proxied requests.
+func newPipelineStatsProxy(transport http.RoundTripper, target *url.URL, key string, tags string) *httputil.ReverseProxy {
+	director := func(req *http.Request) {
+		req.Header.Set("Via", fmt.Sprintf("trace-agent %s", info.Version))
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to the default value
+			// that net/http gives it: Go-http-client/1.1
+			// See https://codereview.appspot.com/7532043
+			req.Header.Set("User-Agent", "")
+		}
+		containerID := req.Header.Get(headerContainerID)
+		if ctags := getContainerTags(containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Additional-Tags", tags)
+		metrics.Count("datadog.trace_agent.pipelines_stats", 1, nil, 1)
+		// URL, Host and key are set in the transport for each outbound request
+	}
+	logger := logutil.NewThrottled(5, 10*time.Second) // limit to 5 messages every 10 seconds
+	return &httputil.ReverseProxy{
+		Director:  director,
+		ErrorLog:  stdlog.New(logger, "pipeline_stats.Proxy: ", 0),
+		Transport: &pipelineStatsTransport{transport, target, key},
```

Comment:
Won't it work just fine without this? The other proxies use a custom transport because they have multiple targets. But here, you're good. You should be able to re-write the request in the Director. Wouldn't that work?

Reply:
Definitely 🤦 Good point, thanks!
Fixed.